ext/native/thin3d/VulkanRenderManager.cpp

```diff
@@ -0,0 +1,1328 @@
+		vkDestroyFence(device, frameData_[i].fence, nullptr);
+	}
+	for (int i = 0; i < ARRAY_SIZE(renderPasses_); i++) {
+		assert(renderPasses_[i] != VK_NULL_HANDLE);
+		vkDestroyRenderPass(device, renderPasses_[i], nullptr);
+	}
+	assert(backbufferRenderPass_ != VK_NULL_HANDLE);
+	vkDestroyRenderPass(device, backbufferRenderPass_, nullptr);
+}
+
+// TODO: Activate this code.
+void VulkanRenderManager::ThreadFunc() {
+	setCurrentThreadName("RenderMan");
+	int threadFrame = -1;  // Increment first, start at 0.
+	while (run_) {
+		{
+			threadFrame++;
+			if (threadFrame >= vulkan_->GetInflightFrames())
+				threadFrame = 0;
+			FrameData &frameData = frameData_[threadFrame];
+			std::unique_lock<std::mutex> lock(frameData.pull_mutex);
+			while (!frameData.readyForRun && run_) {
+				VLOG("PULL: Waiting for frame[%d].readyForRun", threadFrame);
+				frameData.pull_condVar.wait(lock);
+			}
+			VLOG("PULL: frame[%d].readyForRun = false", threadFrame);
+			frameData.readyForRun = false;
+			if (!run_)  // quick exit if bailing.
+				return;
+		}
+		VLOG("PULL: Running frame %d", threadFrame);
+		Run(threadFrame);
+	}
+}
+
+void VulkanRenderManager::BeginFrame() {
+	VkDevice device = vulkan_->GetDevice();
+
+	int curFrame = vulkan_->GetCurFrame();
+	FrameData &frameData = frameData_[curFrame];
+
+	// Make sure the very last command buffer from the frame before the previous has been fully executed.
+	if (useThread) {
+		std::unique_lock<std::mutex> lock(frameData.push_mutex);
+		while (!frameData.readyForFence) {
+			VLOG("PUSH: Waiting for frame[%d].readyForFence = 1", curFrame);
+			frameData.push_condVar.wait(lock);
+		}
+		frameData.readyForFence = false;
+	}
+
+	VLOG("PUSH: Fencing %d", curFrame);
+	vkWaitForFences(device, 1, &frameData.fence, true, UINT64_MAX);
+	vkResetFences(device, 1, &frameData.fence);
+
+	// Must be after the fence - this performs deletes.
+	VLOG("PUSH: BeginFrame %d", curFrame);
+	vulkan_->BeginFrame();
+
+	insideFrame_ = true;
+}
+
+VkCommandBuffer VulkanRenderManager::GetInitCmd() {
+	int curFrame = vulkan_->GetCurFrame();
+	FrameData &frameData = frameData_[curFrame];
+	if (!frameData.hasInitCommands) {
+		VkCommandBufferBeginInfo begin = { VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO };
+		begin.flags = VK_COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT;
+		begin.pInheritanceInfo = nullptr;
+		VkResult res = vkBeginCommandBuffer(frameData.initCmd, &begin);
+		assert(res == VK_SUCCESS);
+		frameData.hasInitCommands = true;
+	}
+	return frameData_[curFrame].initCmd;
+}
+
+void VulkanRenderManager::Sync() {
+
+}
+
+void VulkanRenderManager::BindFramebufferAsRenderTarget(VKRFramebuffer *fb, VKRRenderPassAction color, VKRRenderPassAction depth, uint32_t clearColor, float clearDepth, uint8_t clearStencil) {
+	// Eliminate dupes.
+	if (steps_.size() && steps_.back()->stepType == VKRStepType::RENDER && steps_.back()->render.framebuffer == fb) {
+		if (color != VKRRenderPassAction::CLEAR && depth != VKRRenderPassAction::CLEAR) {
```

Comment:
I wonder if we sometimes set clear unnecessarily and create dupes (e.g. that both clear.)

Since our GL code skips dup binds, which improved perf a little... but probably not common, guess we'll see later.

-[Unknown]